src/modules/model.js

```diff
@@ -78,23 +79,36 @@ class Model {
   }
 
   // eslint-disable-next-line no-underscore-dangle
-  _generateId(data, providedId) {
+  _extractPrimaryKey(data) {
     assert(data instanceof Object && !Array.isArray(data), data);
-    assert(providedId === null || typeof providedId === 'string');
-    if ((typeof providedId !== 'string') === (!Array.isArray(this.primaryKeys))) {
-      throw new MustProvideIdXorPrimaryKeys();
-    }
     if (Array.isArray(this.primaryKeys) && this.primaryKeys.some((k) => data[k] === undefined)) {
       throw new IncompletePrimaryKey();
     }
+    return objectHash(this.primaryKeys.reduce((prev, cur) => Object.assign(prev, { [cur]: data[cur] }), {}));
+  }
+
+  // eslint-disable-next-line no-underscore-dangle
+  _generateId(data, providedId, putOperation = false) {
```

Comment:
Passing an option flag just for validation is usually not a good idea.

Reply:
Rather than passing in the flag I'd always do the check and figure out in the function logic if both are present and only pass the ID.